src/main/kotlin/com/jetbrains/micropython/repl/DeviceCommsManager.kt

```diff
@@ -0,0 +1,97 @@
+package com.jetbrains.micropython.repl
+
+import com.intellij.openapi.components.Service
+import com.intellij.openapi.components.ServiceManager
+import com.intellij.openapi.project.Project
+import com.jediterm.terminal.TtyConnector
+import com.jetbrains.micropython.settings.MicroPythonFacet
+import com.jetbrains.micropython.settings.firstMicroPythonFacet
+import com.pty4j.PtyProcess
+import org.jetbrains.plugins.terminal.LocalTerminalDirectRunner
+import java.lang.ref.WeakReference
+
+sealed class CommsEvent {
+    class ProcessStarted(val ttyConnector: TtyConnector) : CommsEvent()
+    class ProcessCreationFailed(val reason: String) : CommsEvent()
+    object ProcessDestroyed : CommsEvent()
+}
+
+interface CommsEventObserver {
+    fun onCommsEvent(event: CommsEvent)
+}
+
+@Service
+class DeviceCommsManager(project: Project) {
+    private val currentProject = project
+    private var observerRef: WeakReference<CommsEventObserver>? = null
+    private var currentProcess: Process? = null
+    var currentConnector: TtyConnector? = null
+
+    companion object {
+        fun getInstance(project: Project): DeviceCommsManager =
+            ServiceManager.getService(project, DeviceCommsManager::class.java)
+    }
+
+    fun startREPL() {
+        stopREPL()
+
+        currentProject.firstMicroPythonFacet?.let {
+            val devicePath = it.getOrDetectDevicePathSynchronously()
+
+            if (it.pythonPath == null) {
+                notifyObservers(
+                    CommsEvent.ProcessCreationFailed("Valid Python interpreter is needed to start a REPL!")
+                )
+                return
+            }
+
+            if (devicePath == null) {
+                notifyObservers(CommsEvent.ProcessCreationFailed("Device path is not specified, please check settings."))
+                return
+            }
+
+            val terminalRunner = object : LocalTerminalDirectRunner(currentProject) {
+                override fun getInitialCommand(envs: MutableMap<String, String>): MutableList<String> {
+                    return mutableListOf(
+                        it.pythonPath!!,
```

Comment:
You could extract it into a non-null variable above in the outer function and then use it here safely: `val pythonPath = it.pythonPath ?: ...`